docs/design/2019-11-05-index-advisor.md

```diff
@@ -0,0 +1,185 @@
+
+## Proposal
+
+### Input & Output of Index Advisor
+
+The input of index advisor are: `workload`, `parameters`, `statistics`; the output are: `recommend index set`, ` both execution plans before and after adding recommended indexes`. Following shows the details:  
+
+> Input
+1. **Workload**
+
+`workload` can be extracted from *slow_query.log*.
+
+2. **Parameters** (OPTIONAL)
+
+Related parameters and their default values are:
+
+- `database-name="test"`: the database for which index advisor recommend indexes
+- `user="test"`: info required by logging in mysql client
+- `password=""`: info required by logging in mysql client
+- `workload=""`: according to which index advisor recommend indexes
+- `max-exectime=0`: maximum running time of index advisor (unit: minute). Its default value is `0` which means no time constraints for running index advisor
+- `max-index-num=10`: maximum number of indexes recommended by index advisor that is allowed to create 
+- `max-idxcol=1`: maximum number of columns in recommended composite indexes. This proposal set `max-idxcol` as 1
+- `sample-idxstats=false`: whether index advisor is allowed to sample some rows to obtain statistics of composite indexes
+
+3. **Statistics** (OPTIONAL) 
+
+Performance of index advisor greatly depends on the accuracy of SQL optimizer's cost model. Statistics can facilitate SQL optimizer to make a better estimation of `execution time` based on `cost`, which in return improves the performance of index advisor. 
+
+> Output
+
+1. **Recommend index set**
+
+Given input workload, the final recommended indexes for each table. 
+
+2. **Both execution plans of queries in the workload before and after adding recommended indexes** 
+
+This output is mainly for debug purpose, as index advisor is based on underlying SQL optimizer which sometimes works extremely bad. 
+
+----------------------------------------------
+
+### Overview
+
+When recommending suitable indexes for input workload, we should be careful with two constraints: 
+
+1. **Queries in workload cannot be executed**. Apparently, `execution_time` is the most convincing factor which can best explain the efficiency of certain indexes, but in real-world, we cannot afford the price for executing these queries again as executing these queries is slow and resource-intensive. Therefore, we work with `cost` instead to estimate the performance of indexes.
+
+2. **Creating candidate indexes on the table is not allowed**. Creating indexes usually costs a lot, whereas, if the only desired item is `cost` provided by SQL optimizer instead of `execution_time`, we can play some tricks when *creating candidate indexes on tables*. That is, we add these candidate indexes only to the `schema` in memory, which will be accessed by SQL optimizer to produce an execution plan and output its `cost`. 
+
+> Compatibility with TiDB
```

Comment:
Better put it after top layer logic of index advisor. So we can know why we need it.
And maybe we don't even need it and just explain in the `implementation` part.